### ISSUE TITLE
Deprecate reactor methods in tests

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2090]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)| @jml reported|
|-|-|
|Trac ID|trac#2090|
|Type|task|
|Created|2006-09-14 22:22:35Z|

Once trial stops using wait, using any of the following methods will break it:
 * reactor.stop
 * reactor.crash
 * reactor.iterate

Wrap these methods in DeprecationWarnings, at least while tests are running.

<details><summary>Searchable metadata</summary>

```
trac-id__2090 2090
type__task task
reporter__jml jml
priority__highest highest
milestone__Core_2_5 Core-2.5
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__ 
time__1158272555000000 1158272555000000
changetime__1159507055000000 1159507055000000
version__None None
owner__ 
cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml removed owner<br>@jml set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_4"></a>|@radix set owner to @jml|
|-|-|

 * Can you please add a docstring to testHiddenException so I can understand the change that you made with the suppress attribute. Also, if it's testing behavior which is destined for deprecation, a comment explaining that may be useful.

 * Even though those new methods are private, it'd be good if you added docstrings to them.

 * this would be so much better as a reactor wrapper, instead of monkeypatches to the current reactor.

 * Two points: deprecation suppression in the trial internal implementation should be more specific than to suppress all DeprecationWarnings: It should make sure to suppress only those warnings which you really want to suprress. Two: Why don't you just bypass the monkey-patch in these cases so you don't have to worry about the deprecation warning in the first place?

 * _undeprecateReactor should just delete the attribute instead of setting it back, because _deprecateReactor only shadows the original *class* attribute, it doesn't overwrite it.

 * The call should to _deprecateReactor should, perhaps, be moved lower in run()'s implementation, and an extra try/finally should be added surrounding the code between it and _undeprecateReactor.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml set owner to @radix|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_6"></a>|@jml commented|
|-|-|

In reply:

 * Docstrings added
 * `testHiddenException` vastly improved
 * Kept the monkey patches (there was no obvious way to uninstall the reactor)
 * Added a message regex so that only the right warnings are suppressed
 * Can't bypass the monkey patch because some code calls _wait directly (I hate trial). I can't remove that code until I deprecate iterate et al (I hate trial).
 * delattr seems to totally delete the attribute
 * Moved the calls lower, added extra try/finally.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_7"></a>|@jml commented|
|-|-|

I think that glyph was right about adding an option to turn the errors back into warnings. Removing review tag.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_8"></a>|@jml commented|
|-|-|

OK, I've added the option --unclean-warnings. Up for review again.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_9"></a>|@jml commented|
|-|-|

Umm. The last two comments were meant for [#2091](https://github.com/twisted/twisted/issues/2091)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set owner to @jml|
|-|-|

  * Typo in _deprecateReactor docstring - "Depreacte"
  * In general, I think the monkey patch approach here is wrong and will lead to surprising behavior.  Better would be to replace the function object on the class.  However, since this is only for unit tests, and only for supporting now deprecated behavior, and given that it doesn't cause any actual problem in the large suite of tests I tested this branch against, I think it's okay.  If we ever need to do anything like this again though, we should probably do it the right way.

So just fix that docstring and then this looks good to merge.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_11"></a>|@jml set status to `closed`|
|-|-|

(In [18324]) Deprecate direct calls to the reactor from within tests.

 * Author: jml
 * Reviewers: radix, exarkun
 * Fixes [#2090](https://github.com/twisted/twisted/issues/2090)

If Trial ever manages to get rid of _wait, calls to reactor.iterate,
reactor.crash and reactor.stop will cause tests to seriously misbehave
(errors, hanging, crashing etc). This branch deprecates such calls from 
within tests in anticipation of _wait's removal.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
